client/components/PlanningTemplatesModal.tsx

```diff
@@ -0,0 +1,143 @@
+import {ICalendar, IEventTemplate} from '../interfaces';
+import React from 'react';
+import {SearchBar, BoxedListItem, Heading, BoxedList, Modal, TreeSelect, Spacer} from 'superdesk-ui-framework/react';
+import {superdeskApi} from '../superdeskApi';
+
+interface IProps {
+    closeModal: () => void;
+    calendars: Array<ICalendar>;
+    eventTemplates: Array<IEventTemplate>;
+    createEventFromTemplate(template: IEventTemplate): void;
+}
+
+interface IState {
+    activeCalendarFilter?: string;
+    searchQuery: string;
+}
+
+export default class PlanningTemplatesModal extends React.Component<IProps, IState> {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            activeCalendarFilter: null,
+            searchQuery: '',
+        };
+    }
+
+    render(): React.ReactNode {
+        const {gettext} = superdeskApi.localization;
+        const {closeModal, createEventFromTemplate, calendars, eventTemplates} = this.props;
+
+        type ICloseModalProps = {
+            closeModal: () => void;
+        }
+
+        const TemplatesListView: React.FC<ICloseModalProps> = ({closeModal}) => {
+            const calendarFilter = (template: IEventTemplate) => this.state.activeCalendarFilter
+                ? template.data.calendars.map(({qcode}) => qcode).includes(this.state.activeCalendarFilter)
+                : template;
+            /**
+             * Groups the templates by calendar, filters the templates that match the current search query,
+             * filters the templates by the current selected calendar if one is selected,
+             * filters out the groups that don't have any templates
+             */
+            const filteredTemplates = calendars.map((calendar) => ({
```

Comment:
It's not very readable. Maybe it would be cleaner to take templates, filter them, preferrably using 1 filter and only in last step to group it?